internal/controllers/objectsetphases/phase_reconciler.go

```diff
@@ -0,0 +1,482 @@
+	SetControllerReference(owner, obj metav1.Object) error
+}
+
+type adoptionChecker interface {
+	Check(
+		ctx context.Context, owner PhaseObjectOwner, obj client.Object,
+	) (needsAdoption bool, err error)
+}
+
+type patcher interface {
+	Patch(
+		ctx context.Context,
+		desiredObj, currentObj, updatedObj *unstructured.Unstructured,
+	) error
+}
+
+type dynamicCache interface {
+	client.Reader
+	Watch(
+		ctx context.Context, owner client.Object, obj runtime.Object,
+	) error
+}
+
+func NewPhaseReconciler(
+	scheme *runtime.Scheme,
+	writer client.Writer,
+	dynamicCache dynamicCache,
+	ownerStrategy ownerStrategy,
+) *PhaseReconciler {
+	return &PhaseReconciler{
+		scheme:          scheme,
+		writer:          writer,
+		dynamicCache:    dynamicCache,
+		ownerStrategy:   ownerStrategy,
+		adoptionChecker: &defaultAdoptionChecker{ownerStrategy: ownerStrategy},
+		patcher:         &defaultPatcher{writer: writer},
+	}
+}
+
+type PhaseObjectOwner interface {
+	ClientObject() client.Object
+	GetPrevious() []corev1alpha1.PreviousRevisionReference
+	GetStatusRevision() int64
+}
+
+func (r *PhaseReconciler) ReconcilePhase(
+	ctx context.Context, owner PhaseObjectOwner,
+	phase corev1alpha1.ObjectSetTemplatePhase,
+	probe probing.Prober,
+) (failedProbes []string, err error) {
+
+	for _, phaseObject := range phase.Objects {
+		actualObj, err := r.reconcilePhaseObject(ctx, owner, phaseObject)
+		if err != nil {
+			return nil, err
+		}
+
+		if success, message := probe.Probe(actualObj); !success {
+			gvk := actualObj.GroupVersionKind()
+			failedProbes = append(failedProbes,
+				fmt.Sprintf("%s %s %s/%s: %s",
+					gvk.Group, gvk.Kind, actualObj.GetNamespace(), actualObj.GetName(), message))
+		}
+	}
+
+	return
+}
+
+func (r *PhaseReconciler) TeardownPhase(
+	ctx context.Context, owner client.Object,
+	phase corev1alpha1.ObjectSetTemplatePhase,
+) (cleanupDone bool, err error) {
+	var cleanupCounter int
+	objectsToCleanup := len(phase.Objects)
+	for _, phaseObject := range phase.Objects {
+		done, err := r.teardownPhaseObject(ctx, owner, phaseObject)
+		if err != nil {
+			return false, err
+		}
+
+		if done {
+			cleanupCounter++
+		}
+	}
+	return cleanupCounter == objectsToCleanup, nil
+}
+
+func (r *PhaseReconciler) teardownPhaseObject(
+	ctx context.Context, owner client.Object,
+	phaseObject corev1alpha1.ObjectSetObject,
+) (cleanupDone bool, err error) {
+	desiredObj, err := r.desiredObject(ctx, owner, phaseObject)
+	if err != nil {
+		return false, fmt.Errorf("building desired object: %w", err)
+	}
+
+	currentObj := desiredObj.DeepCopy()
+	err = r.dynamicCache.Get(
+		ctx, client.ObjectKeyFromObject(desiredObj), currentObj)
+	if err != nil && errors.IsNotFound(err) {
+		// No matter who the owner of this object is,
+		// it's already gone.
+		return true, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("getting object for teardown: %w", err)
+	}
+
+	if !r.ownerStrategy.IsController(owner, currentObj) {
```

Comment:
Calling `desiredObject` in line 132 sets `owner` as the controller of `desiredObj`. Is it possible the `Get` call in line 138 overwrites this controllership?

Reply:
Nope, the Get in line 138 is just using the key (Name/Namespace) and loads the result into `currentObj`.
`desiredObj` is not changed just read from.

Reply:
Sorry, I don't think I was clear enough. `currentObj` is a deep copy of `desiredObj`, and `owner` is set to be the controller of `desiredObj` in the function `desiredObject` which is called in line 132. I am asking if it is possible for the Get in line 138 to overwrite the controller of `currentObj`, otherwise the condition of this if statement `!r.ownerStrategy.IsController(owner, currentObj` could never be true

Reply:
Ohhhhhhhhhh I think this is what you were addressing in the comment in the test file "// It's super important that we don't check ownership on desiredObj on accident, because that will always return true."

Reply:
Indeed. It's not obvious, that's why the comment is there.